TheKeyOAuth2/TheKeyOAuth2Client+PasswordGrant.swift

```diff
@@ -0,0 +1,126 @@
+                completion(.responseCastingError, nil)
+                return
+            }
+
+            if httpResponse.statusCode == 200, let usableData = data {
+                self.handleSuccessfulPasswordGrant(responseData: usableData, completion: completion)
+                return
+            }
+            
+            if httpResponse.statusCode == 400, let usableData = data {
+                self.handleBadResponse(responseData: usableData, completion: completion)
+                return
+            }
+            
+            if httpResponse.statusCode / 100 == 5 {
+                completion(.serverError, nil)
+                return
+            }
+            
+            completion(.unknownError, nil)
+        }.resume()
+    }
+    
+    private func buildAccessTokenRequest(for username: String, password: String) -> URLRequest? {
+        guard let clientId = clientId, let serverURL = serverURL else {
+            return nil
+        }
+        
+        var formURLString = "username=\(username)"
+        
+        formURLString = formURLString.appending("&password=\(password)")
+        formURLString = formURLString.appending("&client_id=\(clientId)")
+        formURLString = formURLString.appending("&scope=fullticket extended")
+        formURLString = formURLString.appending("&grant_type=password")
+        
+        let tokenURL = serverURL.appendingPathComponent(TheKeyOAuth2TokenEndpoint)
+        
+        var request = URLRequest(url: tokenURL)
+        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
+        request.httpMethod = "POST"
+        request.httpBody = formURLString.data(using: .utf8, allowLossyConversion: false)
+        
+        return request
+    }
+    
+    private func handleSuccessfulPasswordGrant(responseData: Data, completion: @escaping (TheKeyPasswordGrantResult, Error?) -> Void) {
+        do {
+            guard let json = try JSONSerialization.jsonObject(with: responseData, options: .allowFragments) as? Dictionary<String, Any?> else {
+                return
+            }
+
+            setAuthenticationValuesFromJSON(json)
+            saveAuthenticationToKeychain()
+            
+            completion(.success, nil)
+        } catch {
+            completion(.jsonParsingError, error)
+        }
+    }
+    
+    private func handleBadResponse(responseData: Data, completion: @escaping (TheKeyPasswordGrantResult, Error?) -> Void) {
+        do {
+            guard let json = try JSONSerialization.jsonObject(with: responseData, options: .allowFragments) as? Dictionary<String, Any?> else {
+                return
+            }
+            
+            if let errorMessage = json["thekey_authn_error"] as? String, errorMessage == "invalid_credentials" {
+                completion(.badPassword, nil)
+            }
+            
+            // more statuses to come over time..
```

Comment:
current errors:
https://github.com/CruGlobal/thekey/blob/master/support-api/oauth/src/main/java/me/thekey/cas/oauth/Constants.java#L77
https://github.com/CruGlobal/thekey/blob/master/support/oauth-apis/src/main/java/me/thekey/cas/oauth/api/ApiControllerImpl.java#L646

although, there currently isn't anything you can do via an API for any except `invalid_credentials`